odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1671 @@
+
+        For a given functional ``f`` and an element ``translation`` in the
+        domain of ``f``, this operation creates the functional
+        ``f(. - translation)``.
+
+        Parameters
+        ----------
+        translation : `LinearSpaceVector`
+            Element in the domain of the functional
+
+        Returns
+        -------
+        out : `TranslatedFunctional`
+            The functional ``f(. - translation)``
+        """
+        if shift not in self.domain:
+            raise TypeError('the translation {} is not in the domain of the'
+                            'functional {!r}'.format(shift, self))
+
+        return TranslatedFunctional(self, shift)
+
+    def __mul__(self, other):
+        """Return ``self * other``.
+
+        If ``other`` is an operator, this corresponds to composition with the
+        operator:
+
+            ``func * op <==> (x --> func(op(x)))``
+
+        If ``other`` is a scalar, this corresponds to right multiplication of
+        scalars with functionals:
+
+            ``func * scalar <==> (x --> func(scalar * x))``
+
+        If ``other`` is a vector, this corresponds to right multiplication of
+        vectors with functionals:
+
+            ``func * vector <==> (x --> func(vector * x))``
+
+        Note that left and right multiplications are generally different.
+
+        Parameters
+        ----------
+        other : `Operator` or `LinearSpaceVector` or scalar
+            `Operator`:
+            The `Operator.range` of ``other`` must match this functionals
+            `Functional.domain`.
+
+            `LinearSpaceVector`:
+            ``other`` must be an element of this functionals's
+            `Functional.domain`.
+
+            scalar:
+            The `Functional.domain` of this functional must be a
+            `LinearSpace` and ``other`` must be an element of the ``field``
+            of this functional's `Functional.domain`. Note that this
+            ``field`` is also this functional's `Functional.range`.
+
+        Returns
+        -------
+        mul : `Functional`
+            Multiplication result
+
+            If ``other`` is an `Operator`, ``mul`` is a
+            `FunctionalComp`.
+
+            If ``other`` is a scalar, ``mul`` is a
+            `FunctionalRightScalarMult`.
+
+            If ``other`` is a vector, ``mul`` is a
+            `FunctionalRightVectorMult`.
+
+        Notes
+        -------
+        See also `Operator.__mul__`.
+        """
+        if isinstance(other, Operator):
+            return FunctionalComp(self, other)
+        elif isinstance(other, Number):
+            # Left multiplication is more efficient, so we can use this in the
+            # case of linear functional.
+            if self.is_linear:
+                return FunctionalLeftScalarMult(self, other)
+            else:
+                return FunctionalRightScalarMult(self, other)
+        elif isinstance(other, LinearSpaceVector) and other in self.domain:
```

Comment:
I'm sorry, but I'm not sure what to do... delegate the check to what other class and how?